src/passes/CodeFolding.cpp

```diff
@@ -301,6 +304,13 @@ struct CodeFolding : public WalkerPass<ControlFlowWalker<CodeFolding>> {
         // anything exiting that is in all targets is something bad
         return false;
       }
+      // Currently pop instructions are only used for exnref.pop, which is a
+      // pseudo instruction following a catch. In practice, the exnref.pop is a
+      // child of a local.set or a rethrow. We check if the current expression
+      // has a pop child within a limited depth (2).
+      if (containsChild<Pop>(item, 2)) {
```

Comment:
IIUC, this is meant to catch things like this, at a depth of 1?
```
(local.set $x (i32.pop))
```
What happens with say
```
(local.set $x
  (try ..
  (catch
    (local.set $x (i32.pop))
```
In this case there is a more deeply nested `pop`, but it's ok to ignore, I think?

What I'm unsure of is if the 2 here is an optimization, or it's chosen for correctness as we think there is no nesting of >2 that is unsafe, or something like that?

Reply:
It is an optimization. Currently the only toolchain that generates exception code is our LLVM toolchain and it only generates in the form of `(local.set $x (exnref.pop))`. So I think 1 is fine, but 2 is for an extra measure. I don't think we can afford to search all possible children to unlimited depth for every single expression.

The form of code it generates may change after we support values entering blocks, which I'm not sure how will be represented in Binaryen AST. In that case, if we get to generate other forms of code, we may have to change the limit, but for now I think it's fine.

Reply:
So yes, I think moving the entire set in my example is safe. But also that set has a pop at a depth >2 - so it seems to indicate that this isn't just an optimization, as if we ignored the optimization and looked at all depths, we'd not optimize this case.

What concerns me about 2 is that doing a depth of 1, and also when `!NDEBUG` to check for all depths for an assertion, would not work - the assertion would fire. But that is a natural type of assertion to add here, if 2 is not an optimization.

I think we can do probably do something precise and complicated here, but probably that's not worth it. Maybe it's not that bad to check all depths - it's just a linear traversal of each child, and theoretically quadratic with nested blocks, but very low likelihood to have a significant amount of that?

Reply:
We talked about this in person, and decided that limiting depth to 2 is not guaranteed to be safe (Even though the current LLVM wasm toolchain does not generate code that has `pop` in depth more than 1, we can't guarantee that in general, and there can be other toolchains that support wasm EH later). We decided to remove the depth limit, which is overly conservative but at least is correct.